File: _3Sum.ts
function threeSum(nums: number[]): number[][] {


let res = [];

nums.sort((a,b)=> a-b)

for(let i=0; i<nums.length; i++){

    if(i>0 && nums[i]==nums[i-1]){
        continue;
    }

    let l = i+1;
    let r = nums.length -1;
    // console.log('i',i)
    // console.log('l',l)
    // console.log('r',r)

    while(l<r){

        if((nums[i]+nums[l]+nums[r])==0){
            res.push([nums[i],nums[l],nums[r]])
            l++;
            while(nums[l]==nums[l-1] && l<r){
                l++
            }
        }
        else if((nums[i]+nums[l]+nums[r])<0){
            l++;
        }
        else if((nums[i]+nums[l]+nums[r])>0){
            r--;
        }

        
    }
}

return res;

};

File: Python/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        n = len(nums)
        nums.sort()
        
        #-4 -1 -1
        # i
        #   l
        #                  
        for i in range(0,len(nums)-1):
            l = i+1
            r = n-1
            while(l<r):
                if(i==l):
                    l = l+1
                    continue
                if(i==r):
                    r=r-1
                    continue
                
                if(nums[i]+nums[l]+nums[r]==0):
                    temp = []
                    temp.append(nums[i])
                    temp.append(nums[l])
                    temp.append(nums[r])
                    if(temp not in res):
                        res.append(temp)
                    
                    l= l+1
                elif(nums[i]+nums[l]+nums[r]<0):
                    l = l+1
                elif(nums[i]+nums[l]+nums[r]>0):
                    r =r-1
    
        return res
